samples: sensor: Add generic accelerometer polling sample

soburi opened the pull request:
Add a generic polling sample for 3-Axis accelerometers.
This sample demonstrates how to data fetch and print to the console.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

-------------------------------------------

Re-create PR. Previously created one is already close.
https://github.com/zephyrproject-rtos/zephyr/pull/45034

soburi committed:
samples: sensor: Add 3-Axis accelerometer polling sample

Add a polling sample for 3-Axis accelerometers.
This sample demonstrates how to data fetch and print to the console.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

ycsin commented:
I think you need to specify a board and an overlay for it?
https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/sensor/adxl362

galak commented on samples/sensor/adxl345/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Workaround for not being able to have commas in macro arguments
+DT_COMPAT_ADI_ADXL345 := adi,adxl345
```
This should be with the sensor driver and not here.

galak commented on samples/sensor/adxl345/src/main.c:
```diff
@@ -0,0 +1,51 @@
+
+	const struct device *dev = DEVICE_DT_GET_ANY(adi_adxl345);
+
+	if (dev == NULL) {
```
dev will never be null,  replace this check with a `device_is_ready(dev)` check.  See other samples/sensors for example.

galak reviewed (commented):
Looking at main.c, not sure if this adds much beyond what the sensor shell can do.

soburi commented:
@MaureenHelm , @galak 
> This is very close to being a generic accelerometer sample. Can you look at doing something like #47005 to make this work for more than just one accelerometer driver?

> Looking at main.c, not sure if this adds much beyond what the sensor shell can do.


I got it. I'll rewrite this PR as generic  accelerometer sample

soburi committed:
boards: Set devicetree alias for NXP FXOS8700 and compatible nodes

Set accel0 alias for all boards with the FXOS8700 and compatible
accelerometer to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for Bosch BMI270 nodes

Set accel0 alias for all boards with the BMI270 accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for Analog Devices ADXL362 nodes

Set accel0 alias for all boards with the ADXL362 accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Configure I2C/SPI default for boards with accelerometer

For boards with an accelerometer, configure the appropriate bus
(I2C or SPI) default to be enabled if the sensor driver class was enabled.

This change makes no need to add a bunch of board-specific configuration
overlays to the application.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST IIS2DLPC nodes

Set accel0 alias for all boards with the IIS2DLPC accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST LIS2DW12 nodes

Set accel0 alias for all boards with the LIS2DW12 accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST ISM330DHCX nodes

Set accel0 alias for all boards with the ISM330DHCX accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST IIS3DHHC nodes

Set accel0 alias for all boards with the IIS3DHHC accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST LSM6DSO nodes

Set accel0 alias for all boards with the LSM6DSO accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST LSM6DSL nodes

Set accel0 alias for all boards with the LSM6DSL accelerometer
to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST LIS3DH nodes

Set accel0 alias for all boards with the LIS3DH accelerometer
to use by the accel_polling sample.

Add "st,lis2dh" compatibility to resolve dependency
to 96b_wistrio and rak5010_nrf52840.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: Set devicetree alias for ST LIS2DH and compatible nodes

Set accel0 alias for all boards with the LIS2DH and compatible
accelerometer to use by the accel_polling sample.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi commented:
@galak @MaureenHelm 

Rewrite this PR as generic accelerometer sample like as #47005.

galak commented on boards/arm/96b_wistrio/96b_wistrio.dts:
```diff
@@ -93,8 +94,8 @@
-	lis3dh@32 {
-		compatible = "st,lis3dh";
+	lis3dh: lis3dh@32 {
+		compatible = "st,lis3dh", "st,lis2dh";
```
why adding `"st,lis2dh"` here?

soburi replied:
LIS3DH uses LIS2DH driver.
Enabling LIS2DH driver need to set DT_HAS_ST_LIS2DH_ENABLED.
By adding "st,lis2dh" to the compatibile, this option turn on.

If only "st,lis3dh" exists in the compatible option, 
only enable DT_HAS_ST_LIS3DH_ENABLED, but DT_HAS_ST_LIS2DH_ENABLED is not enabled.

galak commented on drivers/sensor/adxl345/Kconfig:
```diff
@@ -3,10 +3,13 @@
 	default y
 	depends on DT_HAS_ADI_ADXL345_ENABLED
 	depends on I2C || SPI
+	default $(dt_compat_enabled,$(DT_COMPAT_ADXL345))
```
Not needed, this is happening with the `default y` and `depends on DT_HAS_ADI_ADXL345_ENABLED`.  Same comment for all the sensor Kconfig changes.

soburi replied:
Removed.

galak reviewed (changes requested):
The driver/sensor/Kconfig changes are not needed and should be removed.

galak commented on boards/arm/bbc_microbit/bbc_microbit.dts:
```diff
@@ -17,6 +17,7 @@
 		sw1 = &buttonB;
 		magn0 = &lsm303agr_magn;
+		accel0 = &mma8653fc;
 		watchdog0 = &wdt0;
```
accel1?

soburi replied:
There are some variations of the BBC microbit. Each has one accelerometer, but the type of sensor is different (LSM303AGR or MMA8653FC).

The dts enables MMA8653FC sensor, and other sensors are disabled.
So, I set mma8653fc to accel0 and no use lsm303agr.

(I think it is intend to enable/disable each sensors by overlay file.)

soburi replied:
There are some variations of the BBC microbit. Each has one accelerometer, but the type of sensor is different (LSM303AGR or MMA8653FC).

The dts enables MMA8653FC sensor, and other sensors are disabled.
So, I set mma8653fc to accel0 and no use lsm303agr.

(I think it is intend to enable/disable each sensors by overlay file.)

galak commented on samples/sensor/accel_polling/sample.yaml:
```diff
@@ -0,0 +1,50 @@
+    harness: console
+    tags: sensors
+    filter: dt_alias_exists("accel0")
+    platform_exclude: sensortile_box thingy53_nrf5340_cpuapp lpcxpresso55s69_cpu1
```
why are these particular platforms excluded?

galak reviewed (commented):
Will leave it to @MaureenHelm if we should remove the samples this replaces in this PR or as a follow up.

gmarull commented on boards/arm/actinius_icarus/Kconfig.defconfig:
```diff
@@ -8,6 +8,13 @@ if BOARD_ACTINIUS_ICARUS || BOARD_ACTINIUS_ICARUS_NS
 config BOARD
 	default "actinius_icarus"
 
+if SENSOR
+
+config I2C
+	default y
+
+endif # SENSOR
```
```
config I2C
	default y if SENSOR
```

soburi replied:
Fixed it.

gmarull commented on boards/arm/actinius_icarus_bee/Kconfig.defconfig:
```diff
@@ -8,6 +8,13 @@ if BOARD_ACTINIUS_ICARUS_BEE || BOARD_ACTINIUS_ICARUS_BEE_NS
 config BOARD
 	default "actinius_icarus_bee"
 
+if SENSOR
+
+config I2C
+	default y
+
+endif # SENSOR
```
ditto (same for all others)

soburi replied:
Fixed it, and also fixes for all similars.

gmarull commented on samples/sensor/accel_polling/README.rst:
```diff
@@ -0,0 +1,38 @@
+.. _accel_polling:
+
+Generic 3-Axis accelerometer polling sample
```
for bisectability: sample should be the last addition

soburi replied:
I reordered commits.

gmarull commented on samples/sensor/accel_polling/sample.yaml:
```diff
@@ -0,0 +1,50 @@
+      - bl5340_dvk                      # lis3dh
+      - b_l4s5i_iot01a                  # lsm6dsl
+
+  sample.sensor.accel_polling.sensortile_box:
+    harness: console
+    tags: sensors
+    platform_allow: sensortile_box
+    harness_config:
+      type: multi_line
+      regex:
+        - "^\\s*lis2dw12@0 \\[m\/s\\^2\\]:    \\(\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*\\)$"
+        - "^\\s*lsm6dso@1 \\[m\/s\\^2\\]:    \\(\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*\\)$"
+        - "^\\s*iis3dhhc@2 \\[m\/s\\^2\\]:    \\(\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*,\\s*-?[0-9\\.]*\\)$"
+    integration_platforms:
+      - sensortile_box                  # lis2dw12, lsm6dso, iisdhhc
```
IMO this overcomplicates the sample.yaml, the generic entry should be sufficient for all cases even if not perfect, in the end we're validating that something is printed...

soburi replied:
Removing board specific cases.

gmarull commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,109 @@
+	SENSOR_CHAN_ACCEL_Z,
+};
+
+void format_sensor_value(struct sensor_value *value, char *buf, size_t len)
+{
+	if (value->val1 == 0) {
+		snprintf(buf, len, "  %c0.%06u", (value->val2 < 0 ? '-' : ' '), abs(value->val2));
+		return;
+	}
+	snprintf(buf, len, "%4d.%06u", value->val1, abs(value->val2));
+}
```
why not use `sensor_value_to_double` and print the value?

soburi replied:
No strong reason is there.
Fix to use sensor_value_to_double.

gmarull commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,109 @@
+			if (ret < 0) {
+				return;
+			}
+			k_sleep(K_MSEC(1000/ARRAY_SIZE(sensors)));
```
why add sleep between different accelerometer reads and not a `k_sleep(K_MSEC(1000))` after all are processed?

soburi replied:
No strong reason is there.
Fix to fetch data from all sensors wit 1s interval.

gmarull commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,109 @@
+#include <stdlib.h>
+#include <stdio.h>
+
+static const struct device *sensors[] = {
+#if DT_NODE_EXISTS(DT_ALIAS(accel0))
+	DEVICE_DT_GET(DT_ALIAS(accel0)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel1))
+	DEVICE_DT_GET(DT_ALIAS(accel1)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel2))
+	DEVICE_DT_GET(DT_ALIAS(accel2)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel3))
+	DEVICE_DT_GET(DT_ALIAS(accel3)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel4))
+	DEVICE_DT_GET(DT_ALIAS(accel4)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel5))
+	DEVICE_DT_GET(DT_ALIAS(accel5)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel6))
+	DEVICE_DT_GET(DT_ALIAS(accel6)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel7))
+	DEVICE_DT_GET(DT_ALIAS(accel7)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel8))
+	DEVICE_DT_GET(DT_ALIAS(accel8)),
+#endif
+#if DT_NODE_EXISTS(DT_ALIAS(accel9))
+	DEVICE_DT_GET(DT_ALIAS(accel9)),
+#endif
+};
```
hardcoding 10 instances this way doesn't look great. First, do we have any board with 10 accelerometers? Second, why not use something like `LISTIFY` to make this more compact?

Maybe a Kconfig option could actually set the number of accelerometers to expect, to be adjusted in a board overlay if > 1.

soburi replied:
Rewrite with LISTIFY macro.  (I didn't know the existence of this macro. It is very helpful!)
As a result of the change, defining an alias for accelX will automatically work with this sample.
I think it's easier than defining a maximum in Kconfig because all you have to do is add an alias in the overlay.

How about this change?

soburi commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,109 @@
+#include <drivers/sensor.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
```
Fixed it.

ycsin commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,69 @@
+				return;
+			}
+		}
+		k_sleep(K_MSEC(1000));
```
could have just used [`k_msleep`](https://docs.zephyrproject.org/2.7.0/reference/kernel/threads/index.html#c.k_msleep)

soburi replied:
Changed it to k_msleep.

soburi commented on samples/sensor/accel_polling/sample.yaml:
```diff
@@ -0,0 +1,27 @@
+    harness: console
+    tags: sensors
+    filter: dt_alias_exists("accel0")
+    platform_exclude: sensortile_box thingy53_nrf5340_cpuapp lpcxpresso55s69_cpu1
```
I forgot delete it.
This line deleted now.

soburi commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,69 @@
+{
+	int ret;
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(sensors); i++) {
```
Fixed it.

soburi commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,69 @@
+	}
+
+	while (1) {
+		for (uint32_t i = 0; i < ARRAY_SIZE(sensors); i++) {
```
Fixed it.

soburi replied:
Thanks for suggestion. Applied it.

soburi commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,69 @@
+#define ACCELEROMETER_DEVICE(i, _)                                                                 \
+	IF_ENABLED(DT_NODE_EXISTS(ACCEL_ALIAS(i)), (DEVICE_DT_GET(ACCEL_ALIAS(i)),))
+
+static const struct device *sensors[] = {LISTIFY(10, ACCELEROMETER_DEVICE, ())};
```
Thanks for suggestion. Applied it.

soburi commented on samples/sensor/accel_polling/README.rst:
```diff
@@ -0,0 +1,38 @@
+This sample requires a 3-Axis accelerometer and needs to set an alias
+that names 'accel0' link to the accelerometer in the devicetree.
+This sample can handle multiple accelerometers by set alias as
+accel1, accel2, ... accel9.
```
Thanks for suggestion. Applied it.

gmarull commented on samples/sensor/accel_polling/sample.yaml:
```diff
@@ -0,0 +1,27 @@
+    harness: console
+    tags: sensors
+    filter: dt_alias_exists("accel0")
+    platform_exclude: lpcxpresso55s69_cpu1
```
why is this required?

soburi replied:
I found correct way to modify dts files about lpcxpresso55s69 board.
Remove this line.

gmarull commented on samples/sensor/accel_polling/src/main.c:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(channels); i++) {
+		ret = sensor_channel_get(dev, channels[i], &accel[i]);
```
`ret` here is ignored?

soburi replied:
Add error check.

gmarull commented on boards/arm/96b_wistrio/Kconfig.defconfig:
```diff
@@ -11,4 +11,7 @@ config BOARD
 	default y if LORA
 
+config I2C
+	default y if SENSOR
```
This pattern is not ideal, but still, it's probably good enough for upstream boards and samples as it minimizes the maintenance work. It should probably be documented together with its limitations.

soburi replied:
> not sure if it is a good idea to default to y as long as SENSOR is enabled, what if user has an overlay to disable the i2c sensor in the default device tree, but has another spi sensor?

My understanding is that the default is should set to what you expect from a usual use case. These boards have sensors on the I2C bus, so this setting should be reasonable.
It may not always be a preferable default value in the complex situations given in the example.
But in such situations, I think it should be explicitly specify configurations, using with prj.conf for examply.

carlescufi commented:
@MaureenHelm ping

MaureenHelm reviewed (approved):
This looks great, thank you!
